File: drills/elixir/solutions/nc115_redundant_connection.exs
defmodule Solution do
  # n nodes and n edges means exactly one cycle. Union-find spots it the moment
  # an edge joins two nodes already connected -- and because the edges are
  # processed in order, the first such edge is the last one that could be
  # removed, which is what the problem asks for.
  def find_redundant_connection(edges) do
    edges
    |> Enum.reduce({%{}, []}, fn [a, b], {parents, found} ->
      root_a = find(parents, a)
      root_b = find(parents, b)

      if root_a == root_b,
        do: {parents, [a, b]},
        else: {Map.put(parents, root_a, root_b), found}
    end)
    |> elem(1)
  end

  defp find(parents, node) do
    case Map.get(parents, node, node) do
      ^node -> node
      parent -> find(parents, parent)
    end
  end
end

File: drills/go/solutions/nc115_redundant_connection.go
package main

func findRedundantConnection(edges [][]int) []int {
	// Union-find over the edges in order: the first edge whose endpoints
	// are already connected closes the cycle, and it is the last such edge
	// in the input among those on the cycle.
	parent := make([]int, len(edges)+1)
	for i := range parent {
		parent[i] = i
	}
	var find func(i int) int
	find = func(i int) int {
		for parent[i] != i {
			parent[i] = parent[parent[i]]
			i = parent[i]
		}
		return i
	}
	for _, edge := range edges {
		a, b := find(edge[0]), find(edge[1])
		if a == b {
			return edge
		}
		parent[a] = b
	}
	return []int{}
}
